### PR TITLE
DEPR: concat ignoring empty objects

### LINKED ISSUE #22186: Concat empty series with mixed TZ doesn't coerce to object
Do we intentionally special case concatenating an empty series with a non-empty series here? I don't think we should. It'd be nice to statically know the output dtype of a `concat` without having to look into the values.

```python
In [2]: import pandas as pd

In [3]: pd.concat([pd.Series([], dtype='datetime64[ns, US/Central]'), pd.Series(['2018'], dtype='datetime64[ns, US/Eastern]')])
Out[3]:
0   2018-01-01 00:00:00-05:00
dtype: datetime64[ns, US/Eastern]

In [4]: pd.concat([pd.Series([], dtype='datetime64[ns, US/Central]'), pd.Series([], dtype='datetime64[ns, US/Eastern]')])
Out[4]: Series([], dtype: object)
```

I'd expect both of those to be object dtype.

### LINKED ISSUE #39122: API: concatting of Series/DataFrame - handling (not skipping) of empty objects
Follow-up on https://github.com/pandas-dev/pandas/pull/38843 and https://github.com/pandas-dev/pandas/pull/39035. 

Currently, we generally (some exceptions can be considered bugs, I think) do not drop empty objects when concatting DataFrames, but we do explicitly drop empties when concatting Series (in `dtypes/concat.py::concat_compat`, for `axis==0`). 

We should make this consistent throughout pandas, and generally I would argue for not skipping empties:  when not skipping empty objects, the resulting dtype of a concat-operation only depends on the input *dtypes*, and not on the exact content (the exact values, how many values (shape)). In general we want to get rid of value-dependent behaviour. In the past we discussed this in the context of the certain values (eg presence of NaNs or not), but I think also the shape should not matter (eg when slicing dataframes before concatting, you can get empties or not depending on *values*).

*If* people agree on going the way of not skipping empties in `concat` (and `append`, and friends), some different areas of work:

* **DataFrames**: in general we already do not skip empty objects when determining the resulting dtype (except for DataFrames with EA columns, see below). However:
  * [ ] Add a bunch of explicit tests for this behaviour.
  * [ ] There are still some bugs, see eg https://github.com/pandas-dev/pandas/issues/32934 about M8[ns] + float64 giving M8[ns] for when empty and object otherwise
* **DataFrame/Series with new (nullable) extension dtypes**: since those are still experimental, I think we can still just make a breaking change here?
  * [ ] Update `concat_compat` to not skip empty nullable EAs
* **Series** (and DataFrame with EA columns): those now skip empties. Can we "just" change this (it's long standing behaviour, and changing it can certainly break code (eg if you at once have no longer a datetime column but object dtype column)? Can we deprecate it? (doesn't sound easy) Or leave it as breaking change for pandas 2.0? (which means keeping the inconsistency for a while longer ..)

So IMO it's mainly the last bullet point (Series/DataFrame with longer-existing EAs) that requires some more discussion on *how* we want to change it.

---

Some illustrative examples:

```python
# dataframe with int64 (default consolidated dtype) and period (EA dtype) column
>>> df1 = pd.DataFrame({'a': [1, 2, 3], 'b': pd.period_range("2012", freq="D", periods=3)})
# dataframe with object dtype columns
>>> df2 = pd.DataFrame({'a': ['a', 'b', 'c'], 'b': ['a', 'b', 'c']})
```

For Series with basic dtype (int64), int64 + object dtype results in object dtype, but not when the object dtype Series is empty:

```python
>>> pd.concat([df1['a'], df2['a']]).dtype
dtype('O')
>>> pd.concat([df1['a'], df2['a'][:0]]).dtype
dtype('int64')
```

For DataFrame, you can see that the int64 + object always gives object (even when one is empty), but for period dtype, the empty object dtype gets ignored:

```python
>>> pd.concat([df1, df2]).dtypes
a    object
b    object
dtype: object
>>> pd.concat([df1, df2[:0]]).dtypes
a       object
b    period[D]
dtype: object
```

cc @pandas-dev/pandas-core

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #39122 (Replace xxxx with the GitHub issue number)
- [x] closes #22186
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jorisvandenbossche

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Rereading some of the previous discussions, would you be OK with start doing this deprecation for empty objects _that are not object/float dtype_? (since those are used as generic "empty" dtype)

**Comment by jbrockmendel** [comment]
> Rereading some of the previous discussions, would you be OK with start doing this deprecation for empty objects that are not object/float dtype? (since those are used as generic "empty" dtype)

I'm not dead-set against that, but would like to push down this road a bit further first.  IIRC the object/float (in particular float) cases were mostly a sticking point in the all-nan cases more than the empty cases.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
looks like something went wrong when merging in the whatsnew

also, this is kind of hard to review, it looks like you're both refactoring and introducing the deprecation warning - any chance to split out the refactoring into a precursor PR please? (or to write a few changes summarising the changes)

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -218,22 +218,29 @@ Deprecations
...
+- Deprecated :func:`concat` behavior when any of the objects being concatenated have length 0; in the past the dtypes of empty objects were ignored when determining the resulting dtype, in a future version they will not (:issue:`39122`)
 - Deprecated :meth:`.DataFrameGroupBy.apply` and methods on the objects returned by :meth:`.DataFrameGroupBy.resample` operating on the grouping column(s); select the columns to operate on after groupby to either explicitly include or exclude the groupings and avoid the ``FutureWarning`` (:issue:`7155`)
 - Deprecated :meth:`.Groupby.all` and :meth:`.GroupBy.any` with datetime64 or :class:`PeriodDtype` values, matching the :class:`Series` and :class:`DataFrame` deprecations (:issue:`34479`)
 - Deprecated :meth:`Categorical.to_list`, use ``obj.tolist()`` instead (:issue:`51254`)
 - Deprecated :meth:`DataFrame._data` and :meth:`Series._data`, use public APIs instead (:issue:`33333`)
 - Deprecated :meth:`DataFrameGroupBy.dtypes`, check ``dtypes`` on the underlying object instead (:issue:`51045`)
+- Deprecated :meth:`DataFrameGroupBy.dtypes`, check ``dtypes`` on the underlying object instead (:issue:`51045`)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply doc/source/whatsnew/v2.1.0.rst:227]: yah, will remove.  for a week or so last month i was getting duplicate entries all the time

**Comment by jbrockmendel** [comment]
> also, this is kind of hard to review, it looks like you're both refactoring and introducing the deprecation warning - any chance to split out the refactoring into a precursor PR please? (or to write a few changes summarising the changes)

I think I can refactor out as a precursor, coming up shortly.

**Comment by jbrockmendel** [comment]
Precursor #53260

File: pandas/core/dtypes/concat.py [hunk 2]
```diff
@@ -38,6 +40,9 @@
     )
 
 
+_dtype_obj = np.dtype(object)
```
Comment (MarcoGorelli) [inline pandas/core/dtypes/concat.py:43]: why is this needed?

Comment (jbrockmendel) [reply pandas/core/dtypes/concat.py:43]: i think its leftover from a previous version.  will remove next time i merge main

File: pandas/core/internals/concat.py [hunk 3]
```diff
@@ -406,16 +408,14 @@ def is_na(self) -> bool:
...
+        """
         blk = self.block
         if blk.dtype.kind == "V":
             return True
-        if not blk._can_hold_na:
-            return False
-
-        values = blk.values
-        if values.size == 0:
-            return True
```
Comment (MarcoGorelli) [inline pandas/core/internals/concat.py:418]: why does this change?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
looks good, but I don't understand the `is_na_without_isna_all` -> `is_na_after_size_and_isna_all_deprecation` change - could you explain please?

File: pandas/core/internals/concat.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/concat.py:418]: Because the future behavior won't depend on `values.size == 0` (note the changed method name/docstring)

Comment (MarcoGorelli) [reply pandas/core/internals/concat.py:418]: sure but the deprecation hasn't been enforced yet, why is this changing already?

Comment (jbrockmendel) [reply pandas/core/internals/concat.py:418]: this method is for checking on the future behavior to see if we need to issue a warning.

**Review by MarcoGorelli — APPROVED** [review]
looks fine to me

@jorisvandenbossche any objections?

**Comment by jbrockmendel** [comment]
@jorisvandenbossche @MarcoGorelli gentle ping

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me

I'd suggest self-merging if it's something you're confident about and have had an approval

**Comment by jorisvandenbossche** [comment]
My apologies for the slow response here, but I would like to get back to this one point (https://github.com/pandas-dev/pandas/pull/52532#issuecomment-1503757602):

> > Rereading some of the previous discussions, would you be OK with start doing this deprecation for empty objects that are not object/float dtype? (since those are used as generic "empty" dtype)
> 
> I'm not dead-set against that, but would like to push down this road a bit further first. IIRC the object/float (in particular float) cases were mostly a sticking point in the all-nan cases more than the empty cases.

I am not fully sure how to interpret the "push down this road a bit further first", but so you would not necessarily be against doing this deprecation only for non-object/float dtypes? 

(I am also happy to do a follow-up PR for that)

**Comment by mroeschke** [comment]
@jbrockmendel when you have a moment, could you also address this test failure (looks like this PR could have used a rebase)

```
FAILED pandas/tests/groupby/test_groupby.py::test_indices_concatenation_order - AssertionError: Did not see expected warning of class 'FutureWarning'
```

**Comment by jbrockmendel** [comment]
> when you have a moment, could you also address this test failure (looks like this PR could have used a rebase)

will take a look now.

> I am not fully sure how to interpret the "push down this road a bit further first", but so you would not necessarily be against doing this deprecation only for non-object/float dtypes?

IIRC at the time of that comment there were other review comments to address (https://github.com/pandas-dev/pandas/pull/52532#discussion_r1163061498) and i wanted to see how "clean" this PR would be once those were addressed.

Also as I said in that comment, my understanding of the motivation to limit to only non-object/float cases was about the all-NA case, not the empty case.  What would the motivation be to special-case the deprecation here?
